商品出品機能

yuricca opened the pull request:
# What
商品出品画面の作成
# Why
商品出品機能実装のため
# Gyazo
- ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/b74cada6bdbebd1e9521273656253ff7
- 価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/df905e50be95e306cd122ec6603a9aa5
- 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/39155e9d186b4eb0771394d24d440dfd
- 入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/08c5f00cd782592b0e54db733c5001b4
- ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/1f1776aa4672a4fbf03ee1e06af3cf45
- テスト結果の画像
https://gyazo.com/9ec856a54ae041f579620d54495dfc11

yuricca committed:
商品出品情報保存のためのモデルとテーブルの作成、ルーティングとコントローラーの設定

yuricca committed:
Active Storageの導入

yuricca committed:
ActiveHashの導入

yuricca committed:
バリデーションの設定

ActiveHashの軽微な修正を含む

yuricca committed:
authenticate_user!の記述

yuricca committed:
商品出品画面の表示と出品情報保存機能の実装

yuricca committed:
モデルの単体テストコードの記述

yuricca committed:
販売手数料と販売利益の計算結果表示

yuricca committed:
Rubocopの実行

div-reviewer commented on app/controllers/application_controller.rb:
```diff
@@ -1,6 +1,7 @@
 class ApplicationController < ActionController::Base
   before_action :basic_auth
   before_action :configure_permitted_parameters, if: :devise_controller?
+  before_action :authenticate_user!, except: :index
```
```ruby
before_action :authenticate_user!, except: :index
```

ログインされていなければ〜の実装ありがとうございました。

**結論、こちらは商品のコントローラーに記載してあげましょう。**

理由は、application_controllerに記述してしまうと、今後実装する全てのコントローラーで現在記載して頂いているコードが適用されてしまうためです。

div-reviewer commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,24 @@
+  end
+
+  def create
+    @item = Item.create(item_params)
```
```ruby
@item = Item.create(item_params)
```
こちらはcreateメソッドではなくnewメソッドを用いて

```ruby
@item = Item.new(item_params)
```
と記述しましょう。

createメソッドを使用するとデータベースに商品データが保存されてしまうためです。
１行下でも@item.saveの記述によって保存処理が行われるため、データベースに二重でデータが保存されてしまいます。

div-reviewer commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,78 @@
+        @item.price = '１２３４５６'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not a number')
+      end
```
テストのit文もとても丁寧に記述して下さりありがとうございます！
第3者からみて見やすいコードがばっちり書けていますね。

お手数ですが以下のテストも追加しましょう！
```ruby
【ユーザー系】
userが紐付いていなければ出品できない
```

div-reviewer reviewed (commented):
yuricca様、学習お疲れ様です！
お忙しい中商品出品の実装、レビュー依頼ありがとうございました。

全体的にコードがとても綺麗に書けているかと思われます。
今回はテストの追加とコントローラーでの修正をあわせて3点程お願いしております。
お手数ですがご確認頂けますでしょうか？

実装が難しい場合、疑問や質問がございましたらぜひコネクトメンターにご相談下さい。
完成まであともう少しかと思われます。最後まで頑張っていきましょう！

yuricca committed:
コードレビューを受けての修正

①authenticate_user! の記述場所をapplication_controller.rb → items_controller.rbに修正
②createメソッドの修正
③テストの追加

yuricca commented:
とても丁寧にレビューをしていただきありがとうございます。
コードも褒めてくださりモチベーションが上がります！
ご指摘いただいた点を修正しましたので、ご確認をお願いいたします。
# Gyazo
- 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/e20f75ed8b666368a32572915a22fb53
- ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/dc9ef4da1a5d488e668f18849c13043c
- テスト結果の画像
https://gyazo.com/e70179c028ab63e35ae91c7274c1e5d0

div-reviewer reviewed (commented):
yuricca様、学習お疲れ様です！
お忙しい中直ぐの修正と再レビュー依頼ありがとうございました。

>とても丁寧にレビューをしていただきありがとうございます。
コードも褒めてくださりモチベーションが上がります！

こちらこそ、お忙しい中での前向きな実装と、
お心遣いのこもったお言葉ありがとうございました！！とても嬉しいです。

前回コメントしていた箇所もご丁寧に確認と修正して頂きありがとうございました。
欲しい機能が実装できているかと思われます。
よって、こちら`LGTM`とさせていただきます！実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！

次は「商品一覧機能」かと思われます！引き続き頑張りましょう！応援しております！
